pkg/pool-weighted/test/LiquidityBootstrappingPool.test.ts

```diff
@@ -1,2 +1,161 @@
+    });
+
+    it('LP cannot pause swaps', async () => {
+      await expect(pool.setPublicSwap(lp, false)).to.be.revertedWith('SENDER_NOT_ALLOWED');
+    });
+  });
+
+  describe('joins', () => {
+    sharedBeforeEach('initialize pool', async () => {
+      await pool.init({ recipient: owner, initialBalances });
+
+      //const action = await actionId(pool.instance, '_onJoinPool');
+      //await pool.vault.grantRole(action, owner);
+    });
+
+    it('no public LPs', async () => {
+      await expect(pool.joinGivenIn({ recipient: lp, amountsIn: initialBalances })).to.be.revertedWith(
+        'SENDER_NOT_ALLOWED'
+      );
+    });
+
+    it('allows owner to join', async () => {
+      await expect(pool.joinGivenIn({ recipient: owner, from: owner, amountsIn: initialBalances })).to.not.be.reverted;
+    });
+  });
+
+  describe('update weights', () => {
+    sharedBeforeEach('deploy tokens', async () => {
+      const action = await actionId(pool.instance, 'updateWeightsGradually');
+      await pool.vault.grantRole(action, owner);
+    });
+
+    it('only allows owner to update weights', async () => {
+      const now = await currentTimestamp();
+
+      await expect(pool.updateWeightsGradually(lp, now, now, weights)).to.be.revertedWith('SENDER_NOT_ALLOWED');
+    });
+
+    context('call update weights', () => {
+      // startWeights = [fp(0.3), fp(0.55), fp(0.1), fp(0.5)];
+      const endWeights = [fp(0.15), fp(0.25), fp(0.55), fp(0.05)];
+      const halfWeights = [fp(0.225), fp(0.4), fp(0.325), fp(0.05)];
+      let now, startTime: BigNumber, endTime: BigNumber;
+
+      sharedBeforeEach('updateWeightsGradually', async () => {
+        now = await currentTimestamp();
+        startTime = now.add(MINUTE * 10);
+        endTime = startTime.add(MINUTE * 60);
+
+        await pool.updateWeightsGradually(owner, startTime, endTime, endWeights);
+      });
+
+      it('stores the params', async () => {
+        const updateParams = await pool.getGradualUpdateParams();
+
+        expect(updateParams.startTime).to.equalWithError(startTime, 0.001);
+        expect(updateParams.endTime).to.equalWithError(endTime, 0.001);
+        expect(updateParams.endWeights).to.equalWithError(endWeights, 0.001);
+      });
+
+      it('gets start weights if called early', async () => {
+        const normalizedWeights = await pool.getNormalizedWeights();
+
+        // Need to decrease precision
+        expect(normalizedWeights).to.equalWithError(pool.normalizedWeights, 0.0001);
+      });
+
+      it('gets end weights if called late', async () => {
+        await advanceTime(MINUTE * 100);
+        const normalizedWeights = await pool.getNormalizedWeights();
+
+        // Need to decrease precision
+        expect(normalizedWeights).to.equalWithError(endWeights, 0.0001);
+      });
+
+      it('gets intermediate weights if called halfway through', async () => {
+        await advanceTime(MINUTE * 40);
```

Comment:
Same here, it is not obvious that this is the halfway point. 

Other test scenarios we should have:
 - some more percentages (e.g. 30 and 70): 50% is a particularly tricky one because it doesn't detect errors where you're taking the wrong end
 - calling 'start update' after start time

Reply:
Ok; the tests are in better shape. Will look into generalizing further (to different counts and weights) later